File: common/src/main/java/com/stardust/pio/TraditionalFileProvider.java
package com.stardust.pio;

import android.util.Log;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * 传统 File API 实现
 * 适用于拥有完全存储权限或应用私有目录的场景
 */
public class TraditionalFileProvider implements IFileProvider {

    private static final String TAG = "TraditionalFileProvider";

    private String mWorkingDirectory;

    public TraditionalFileProvider() {
        mWorkingDirectory = "/";
        Log.d(TAG, "Created: workingDir=/");
    }

    public TraditionalFileProvider(String workingDirectory) {
        mWorkingDirectory = workingDirectory;
        Log.d(TAG, "Created: workingDir=" + workingDirectory);
    }

    @Override
    public boolean exists(String path) {
        return new File(resolvePath(path)).exists();
    }

    @Override
    public boolean isFile(String path) {
        return new File(resolvePath(path)).isFile();
    }

    @Override
    public boolean isDirectory(String path) {
        return new File(resolvePath(path)).isDirectory();
    }

    @Override
    public boolean mkdir(String path) {
        return new File(resolvePath(path)).mkdir();
    }

    @Override
    public boolean mkdirs(String path) {
        return new File(resolvePath(path)).mkdirs();
    }

    @Override
    public boolean delete(String path) {
        return new File(resolvePath(path)).delete();
    }

    @Override
    public boolean deleteRecursively(String path) {
        File file = new File(resolvePath(path));
        return deleteRecursive(file);
    }

    private boolean deleteRecursive(File file) {
        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    deleteRecursive(child);
                }
            }
        }
        return file.delete();
    }

    @Override
    public boolean rename(String path, String newName) {
        File file = new File(resolvePath(path));
        File newFile = new File(file.getParent(), newName);
        return file.renameTo(newFile);
    }

    @Override
    public boolean move(String fromPath, String toPath) {
        File from = new File(resolvePath(fromPath));
        File to = new File(resolvePath(toPath));
        return from.renameTo(to);
    }

    @Override
    public boolean copy(String fromPath, String toPath) {
        try {
            File from = new File(resolvePath(fromPath));
            File to = new File(resolvePath(toPath));
            
            if (from.isDirectory()) {
                return copyDirectory(from, to);
            } else {
                return copyFile(from, to);
            }
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    private boolean copyFile(File from, File to) throws Exception {
        try (FileInputStream fis = new FileInputStream(from);
             FileOutputStream fos = new FileOutputStream(to)) {
            byte[] buffer = new byte[8192];
            int len;
            while ((len = fis.read(buffer)) > 0) {
                fos.write(buffer, 0, len);
            }
            return true;
        }
    }

    private boolean copyDirectory(File from, File to) throws Exception {
        if (!to.exists()) {
            to.mkdirs();
        }
        
        File[] children = from.listFiles();
        if (children == null) return true;
        
        for (File child : children) {
            File newChild = new File(to, child.getName());
            if (child.isDirectory()) {
                copyDirectory(child, newChild);
            } else {
                copyFile(child, newChild);
            }
        }
        return true;
    }

    @Override
    public List<FileInfo> listFiles(String path) {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "listFiles: path=" + path + ", resolved=" + resolvedPath);
        List<FileInfo> result = new ArrayList<>();
        File dir = new File(resolvedPath);
        File[] files = dir.listFiles();
        if (files == null) {
            Log.w(TAG, "listFiles: listFiles() returned null, path may not be a directory or no permission");
            return result;
        }
        
        for (File file : files) {
            result.add(new FileInfo(
                    file.getName(),
                    file.getAbsolutePath(),
                    file.isDirectory(),
                    file.length(),
                    file.lastModified()
            ));
        }
        Log.d(TAG, "listFiles: found " + result.size() + " items");
        return result;
    }

    @Override
    public String read(String path, String encoding) {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "read: path=" + path + ", resolved=" + resolvedPath + ", encoding=" + encoding);
        try {
            String result = PFiles.read(resolvedPath, encoding);
            Log.d(TAG, "read: success, length=" + (result != null ? result.length() : 0));
            return result;
        } catch (Exception e) {
            Log.e(TAG, "read: error=" + e.getMessage(), e);
            return null;
        }
    }

    @Override
    public String read(String path) {
        return read(resolvePath(path), "UTF-8");
    }

    @Override
    public byte[] readBytes(String path) {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "readBytes: path=" + path + ", resolved=" + resolvedPath);
        try {
            File file = new File(resolvedPath);
            try (FileInputStream fis = new FileInputStream(file)) {
                byte[] data = new byte[(int) file.length()];
                fis.read(data);
                Log.d(TAG, "readBytes: success, size=" + data.length);
                return data;
            }
        } catch (Exception e) {
            Log.e(TAG, "readBytes: error=" + e.getMessage(), e);
            return null;
        }
    }

    @Override
    public InputStream openInputStream(String path) throws Exception {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "openInputStream: resolved=" + resolvedPath);
        return new BufferedInputStream(new FileInputStream(resolvedPath));
    }

    @Override
    public boolean write(String path, String content, String encoding) {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "write: path=" + path + ", resolved=" + resolvedPath + ", contentLen=" + content.length());
        try {
            PFiles.write(resolvedPath, content, encoding);
            Log.d(TAG, "write: success");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "write: error=" + e.getMessage(), e);
            return false;
        }
    }

    @Override
    public boolean write(String path, String content) {
        return write(resolvePath(path), content, "UTF-8");
    }

    @Override
    public boolean append(String path, String content, String encoding) {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "append: resolved=" + resolvedPath);
        try {
            PFiles.append(resolvedPath, content, encoding);
            return true;
        } catch (Exception e) {
            Log.e(TAG, "append: error=" + e.getMessage(), e);
            return false;
        }
    }

    @Override
    public boolean writeBytes(String path, byte[] bytes) {
        String resolvedPath = resolvePath(path);
        Log.d(TAG, "writeBytes: resolved=" + resolvedPath + ", size=" + bytes.length);
        try (FileOutputStream fos = new FileOutputStream(resolvedPath)) {
            fos.write(bytes);
            Log.d(TAG, "writeBytes: success");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "writeBytes: error=" + e.getMessage(), e);
            return false;
        }
    }

    @Override
    public OutputStream openOutputStream(String path) throws Exception {
        return openOutputStream(path, false);
    }

    @Override
    public OutputStream openOutputStream(String path, boolean append) throws Exception {
        return new BufferedOutputStream(new FileOutputStream(resolvePath(path), append));
    }

    @Override
    public long length(String path) {
        return new File(resolvePath(path)).length();
    }

    @Override
    public long lastModified(String path) {
        return new File(resolvePath(path)).lastModified();
    }

    @Override
    public String getName(String path) {
        return new File(resolvePath(path)).getName();
    }

    @Override
    public String getParent(String path) {
        return new File(resolvePath(path)).getParent();
    }

    @Override
    public String getExtension(String path) {
        return PFiles.getExtension(resolvePath(path));
    }

    @Override
    public boolean isAccessible(String path) {
        // 传统模式下，只要文件存在或父目录可写就认为可访问
        String resolvedPath = resolvePath(path);
        File file = new File(resolvedPath);
        boolean result;
        if (file.exists()) {
            result = file.canRead();
            Log.d(TAG, "isAccessible: path=" + resolvedPath + ", exists=true, canRead=" + result);
        } else {
            File parent = file.getParentFile();
            result = parent != null && parent.canWrite();
            Log.d(TAG, "isAccessible: path=" + resolvedPath + ", exists=false, parentWritable=" + result);
        }
        return result;
    }

    @Override
    public String getWorkingDirectory() {
        return mWorkingDirectory;
    }

    @Override
    public void setWorkingDirectory(String path) {
        Log.d(TAG, "setWorkingDirectory: " + path);
        mWorkingDirectory = path;
    }

    @Override
    public String resolvePath(String path) {
        if (path == null || path.isEmpty()) {
            return mWorkingDirectory;
        }
        
        File file = new File(path);
        if (file.isAbsolute()) {
            return path;
        }
        
        return new File(mWorkingDirectory, path).getAbsolutePath();
    }
}

File: common/src/main/java/com/stardust/pio/TraditionalFileProvider.kt
package com.stardust.pio

import android.util.Log
import java.io.*

/**
 * 传统 File API 实现
 * 适用于拥有完全存储权限或应用私有目录的场景
 */
class TraditionalFileProvider : IFileProvider {

    companion object {
        private const val TAG = "TraditionalFileProvider"
    }

    private var workingDirectory: String

    constructor() {
        workingDirectory = "/"
        Log.d(TAG, "Created: workingDir=/")
    }

    constructor(workingDirectory: String) {
        this.workingDirectory = workingDirectory
        Log.d(TAG, "Created: workingDir=$workingDirectory")
    }

    override fun exists(path: String): Boolean {
        val resolved = resolvePath(path)
        val result = File(resolved).exists()
        Log.d(TAG, "exists: path=$path, resolved=$resolved, result=$result")
        return result
    }

    override fun isFile(path: String): Boolean {
        val resolved = resolvePath(path)
        return File(resolved).isFile
    }

    override fun isDirectory(path: String): Boolean {
        val resolved = resolvePath(path)
        return File(resolved).isDirectory
    }

    override fun mkdir(path: String): Boolean {
        val resolved = resolvePath(path)
        return File(resolved).mkdir()
    }

    override fun mkdirs(path: String): Boolean {
        val resolved = resolvePath(path)
        return File(resolved).mkdirs()
    }

    override fun delete(path: String): Boolean {
        val resolved = resolvePath(path)
        return File(resolved).delete()
    }

    override fun deleteRecursively(path: String): Boolean {
        val resolved = resolvePath(path)
        return deleteRecursive(File(resolved))
    }

    private fun deleteRecursive(file: File): Boolean {
        if (file.isDirectory) {
            file.listFiles()?.forEach { deleteRecursive(it) }
        }
        return file.delete()
    }

    override fun rename(path: String, newName: String): Boolean {
        val resolved = resolvePath(path)
        val file = File(resolved)
        val newFile = File(file.parent, newName)
        return file.renameTo(newFile)
    }

    override fun move(fromPath: String, toPath: String): Boolean {
        val fromResolved = resolvePath(fromPath)
        val toResolved = resolvePath(toPath)
        return File(fromResolved).renameTo(File(toResolved))
    }

    override fun copy(fromPath: String, toPath: String): Boolean {
        val fromResolved = resolvePath(fromPath)
        val toResolved = resolvePath(toPath)
        return try {
            val from = File(fromResolved)
            val to = File(toResolved)
            if (from.isDirectory) copyDirectory(from, to) else copyFile(from, to)
        } catch (e: Exception) {
            Log.e(TAG, "copy: error=${e.message}", e)
            false
        }
    }

    private fun copyFile(from: File, to: File): Boolean {
        FileInputStream(from).use { fis ->
            FileOutputStream(to).use { fos ->
                val buffer = ByteArray(8192)
                var len: Int
                while (fis.read(buffer).also { len = it } > 0) {
                    fos.write(buffer, 0, len)
                }
            }
        }
        return true
    }

    private fun copyDirectory(from: File, to: File): Boolean {
        if (!to.exists()) to.mkdirs()
        from.listFiles()?.forEach { child ->
            val newChild = File(to, child.name)
            if (child.isDirectory) copyDirectory(child, newChild) else copyFile(child, newChild)
        }
        return true
    }

    override fun listFiles(path: String): List<IFileProvider.FileInfo> {
        val resolvedPath = resolvePath(path)
        val result = mutableListOf<IFileProvider.FileInfo>()
        val dir = File(resolvedPath)
        val files = dir.listFiles() ?: return result
        files.forEach { file ->
            val mimeType = guessMimeType(file)
            val flags = computeFileFlags(file)
            result.add(IFileProvider.FileInfo(
                file.name,
                file.absolutePath,
                file.isDirectory,
                file.length(),
                file.lastModified(),
                mimeType,
                flags
            ))
        }
        return result
    }

    override fun getMimeType(path: String): String? {
        val resolved = resolvePath(path)
        val file = File(resolved)
        if (!file.exists()) return null
        return guessMimeType(file)
    }

    /**
     * 根据 File 属性计算 flags
     */
    private fun computeFileFlags(file: File): Int {
        var flags = 0
        if (file.canWrite()) {
            flags = flags or IFileProvider.FileFlags.SUPPORTS_WRITE
        }
        // 传统 File API 通常支持删除、重命名、复制、移动（如果有权限）
        val parent = file.parentFile
        if (parent != null && parent.canWrite()) {
            flags = flags or IFileProvider.FileFlags.SUPPORTS_DELETE
            flags = flags or IFileProvider.FileFlags.SUPPORTS_RENAME
            flags = flags or IFileProvider.FileFlags.SUPPORTS_MOVE
        }
        // 复制只需要读取权限
        if (file.canRead()) {
            flags = flags or IFileProvider.FileFlags.SUPPORTS_COPY
        }
        // 传统文件系统不是虚拟文档
        // 缩略图：图片文件支持
        val mimeType = guessMimeType(file)
        if (mimeType != null && mimeType.startsWith("image/")) {
            flags = flags or IFileProvider.FileFlags.SUPPORTS_THUMBNAIL
        }
        return flags
    }

    /**
     * 猜测 MIME 类型
     */
    private fun guessMimeType(file: File): String? {
        if (file.isDirectory) {
            return "application/vnd.android.document/directory"
        }
        val name = file.name
        val lastDot = name.lastIndexOf('.')
        if (lastDot < 0 || lastDot >= name.length - 1) {
            return "application/octet-stream"
        }
        val extension = name.substring(lastDot + 1).lowercase()
        return when (extension) {
            // 文本类型
            "txt" -> "text/plain"
            "html", "htm" -> "text/html"
            "css" -> "text/css"
            "js" -> "application/javascript"
            "json" -> "application/json"
            "xml" -> "application/xml"
            "md" -> "text/markdown"
            // 图片类型
            "jpg", "jpeg" -> "image/jpeg"
            "png" -> "image/png"
            "gif" -> "image/gif"
            "webp" -> "image/webp"
            "svg" -> "image/svg+xml"
            "bmp" -> "image/bmp"
            // 音频类型
            "mp3" -> "audio/mpeg"
            "wav" -> "audio/wav"
            "ogg" -> "audio/ogg"
            "m4a" -> "audio/mp4"
            // 视频类型
            "mp4" -> "video/mp4"
            "mkv" -> "video/x-matroska"
            "webm" -> "video/webm"
            "avi" -> "video/x-msvideo"
            // 压缩文件
            "zip" -> "application/zip"
            "gz" -> "application/gzip"
            "rar" -> "application/vnd.rar"
            "7z" -> "application/x-7z-compressed"
            // 文档类型
            "pdf" -> "application/pdf"
            "doc" -> "application/msword"
            "docx" -> "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            "xls" -> "application/vnd.ms-excel"
            "xlsx" -> "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            "ppt" -> "application/vnd.ms-powerpoint"
            "pptx" -> "application/vnd.openxmlformats-officedocument.presentationml.presentation"
            // Android 特有
            "apk" -> "application/vnd.android.package-archive"
            // Java
            "java" -> "text/x-java-source"
            "kt" -> "text/x-kotlin"
            // 其他
            else -> "application/octet-stream"
        }
    }

    override fun read(path: String, encoding: String): String? {
        return try {
            PFiles.read(resolvePath(path), encoding)
        } catch (e: Exception) {
            Log.e(TAG, "read: error=${e.message}", e)
            null
        }
    }

    override fun read(path: String): String? = read(resolvePath(path), "UTF-8")

    override fun readBytes(path: String): ByteArray? {
        return try {
            File(resolvePath(path)).readBytes()
        } catch (e: Exception) {
            Log.e(TAG, "readBytes: error=${e.message}", e)
            null
        }
    }

    override fun openInputStream(path: String): InputStream {
        return BufferedInputStream(FileInputStream(resolvePath(path)))
    }

    override fun write(path: String, content: String, encoding: String): Boolean {
        return try {
            PFiles.write(resolvePath(path), content, encoding)
            true
        } catch (e: Exception) {
            Log.e(TAG, "write: error=${e.message}", e)
            false
        }
    }

    override fun write(path: String, content: String): Boolean = write(resolvePath(path), content, "UTF-8")

    override fun append(path: String, content: String, encoding: String): Boolean {
        return try {
            PFiles.append(resolvePath(path), content, encoding)
            true
        } catch (e: Exception) {
            Log.e(TAG, "append: error=${e.message}", e)
            false
        }
    }

    override fun append(path: String, content: String): Boolean = append(path, content, "UTF-8")

    override fun writeBytes(path: String, bytes: ByteArray): Boolean {
        return try {
            File(resolvePath(path)).writeBytes(bytes)
            true
        } catch (e: Exception) {
            Log.e(TAG, "writeBytes: error=${e.message}", e)
            false
        }
    }

    override fun openOutputStream(path: String): OutputStream = openOutputStream(path, false)

    override fun openOutputStream(path: String, append: Boolean): OutputStream {
        return BufferedOutputStream(FileOutputStream(resolvePath(path), append))
    }

    override fun length(path: String): Long = File(resolvePath(path)).length()

    override fun lastModified(path: String): Long = File(resolvePath(path)).lastModified()

    override fun getName(path: String): String = File(resolvePath(path)).name

    override fun getParent(path: String): String? = File(resolvePath(path)).parent

    override fun getExtension(path: String): String = PFiles.getExtension(resolvePath(path))

    override fun isAccessible(path: String): Boolean {
        val resolvedPath = resolvePath(path)
        val file = File(resolvedPath)
        return if (file.exists()) file.canRead() else file.parentFile?.canWrite() ?: false
    }

    override fun getWorkingDirectory(): String = workingDirectory

    override fun setWorkingDirectory(path: String) {
        Log.d(TAG, "setWorkingDirectory: $path")
        workingDirectory = path
    }

    override fun resolvePath(path: String): String {
        if (path.isEmpty()) return workingDirectory
        val file = File(path)
        return if (file.isAbsolute) path else File(workingDirectory, path).absolutePath
    }
}
